Propose an exception handling criteria for Silver integration quality (#1344)
home-assistant/developers.home-assistant: docs/integration_quality_scale_index.md

- Handles expiration of auth credentials. Refresh if possible or print correct error and fail setup. If based on a config entry, should trigger a new config entry flow to re-authorize.  ([docs](config_entries_config_flow_handler.md#reauthentication))
- Handles internet unavailable. Log a warning once when unavailable, log once when reconnected.
- Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.

> - Operations that fail have propper exception handling e.g. a service call or entity method *Set HVAC Mode* failure is thrown as `HomeAssistantError`

Reviewer: If the user input is invalid we should raise `ValueError` otherwise we should raise `HomeAssistantError`.

Author: Re-written to incorporate user input failures.

> - Operations like a service call or entity method (e.g. *Set HVAC Mode*) have proper exception handling. If user input is invalid raise `ValueErorr`, otherwise raise `HomeAssistantError`.

Reviewer 2: ```suggestion
- Operations like service calls and entity methods (e.g. *Set HVAC Mode*) have proper exception handling. If user input is invalid raise `ValueError`, otherwise raise `HomeAssistantError`.
```
I would like to have a better wording for this, as it leaves me to wonder _how_ to handle different types of exceptions. Is `ValueError` only to be raised on invalid input, and rest of the exception should be wrapped inside `HomeAssistantError` or something else? That is, should a common pattern after this change on service calls be something like this:
```
def some_service_call(self, input: int):
    if input < 0:
        raise ValueError("invalid input")
    try:
        self.device_ref.execute_some_service_call(input)
    except Exception as ex:
        raise HomeAssistantError() from ex
```
which I think feels clumsy.

Shortly put, it'd be great to have a description on _when_ `HomeAssistantError` is to be raised instead of bubbling it upwards :-)

Author: There are other rules like "Never catch `Exception`"  that also come in to play, but otherwise your example seems close to what I think we're saying. e.g. catching a communication error raised by the device, instead of letting it fall through it needs to be raised as `HomeAssistantError`. I guess the point isn't just to wrap things arbitrarily, but explicitly to raise exceptions thrown by devices in a way that home assistant understands.

1) Do you agree with that standard?
2)Then if we're agreed,  then open to suggestions to improve the wording.

Reviewer 2: Would the wording be something like "All user actionable exceptions should be raised as instances `HomeAssistantError` and contain reasonable error messages", assuming that these are/could be something that are displayed to the user directly?

I shouldn't have written `Exception` there, the idea was only to confirm my understanding of the idea that upstream exceptions should be captured and re-raised.

Reviewer 2: How about "`ValueError` should be raised on invalid user input, otherwise all upstream exceptions should be wrapped inside `HomeAssistantError` (because of reasons)"? That still reads like a mouthful, but it would be much clearer.

The point I'm trying to make is that using explicit wording is better than vague expressions.

Author: Thanks. It's helpful to understand what part seemed vague. How's this?\

```
Operations like service calls and entity methods (e.g. *Set HVAC Mode*) have proper exception handling. Raise `ValueError` on invalid user input and raise `HomeAssistantError` for other failures such as a problem communicating with a device.
```

- Handles expiration of auth credentials. Refresh if possible or print correct error and fail setup. If based on a config entry, should trigger a new config entry flow to re-authorize.  ([docs](config_entries_config_flow_handler.md#reauthentication))
- Handles internet unavailable. Log a warning once when unavailable, log once when reconnected.
- Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.

> - Operations like a service call or entity method (e.g. *Set HVAC Mode*) have propper exception handling. If user input is invalid raise `ValueErorr`, otherwise raise `HomeAssistantError`.

Reviewer: ```suggestion
- Operations like a service call or entity method (e.g. *Set HVAC Mode*) have proper exception handling. If user input is invalid raise `ValueErorr`, otherwise raise `HomeAssistantError`.
```
